### PR TITLE
feat(cloner): clone strategies

### PR DESCRIPTION
**Pull request by sgaist**
## Describe your changes

This PR implements strategies about acting before cloning.

There are three strategies:
- default: do as usual and git clone
- notifexists: don't clone if the target folder already exists
- overwrite: delete the target folder before cloning

This happens before trying to clone anything.

The goal is to allow restarting the session which means also the init container and not fail the cloning if it already happened before.

As a drive by cleanup, the `.git` extension is removed if part of the URL as happen when using the git command.

## Issue ticket number and link

Part of: #612
Part of https://github.com/SwissDataScienceCenter/renku/issues/3679

### CODE REVIEW

**Review by olevski — CHANGES_REQUESTED** [review]
One question about the "default" strategy. All else looks good.

File: cloner/cmd/clone.go [hunk 1]
```diff
@@ -32,18 +33,21 @@ import (
...
-	remote     string
-	revision   string
-	path       string
-	verbose    bool
+	configPath       string
+	remote           string
+	revision         string
+	path             string
+	verbose          bool
+	CloningStratgies []string = []string{"notifexist", "overwrite", "default"}
```
[the remark this answers is not in the document]
Comment (sgaist) [reply cloner/cmd/clone.go:49]: The `default` value means "no pre-processing" before cloning so if the target folder exists, the clone command will abort.

The other two will trigger checks and actions before the cloning happens.

File: cloner/cmd/clone.go [hunk 2]
```diff
@@ -78,20 +84,44 @@ func clone(cmd *cobra.Command, args []string) {
 
 	endpoint, err := transport.NewEndpoint(remote)
 	if err != nil {
-		log.Fatal("failed to parse remote", err)
+		log.Fatal("failed to parse remote: ", err)
 	}
 
 	splittedRepo := strings.FieldsFunc(endpoint.Path, func(c rune) bool { return c == '/' }) // FieldsFunc handles repeated and beginning/ending separator characters more sanely than Split
 	if len(splittedRepo) < 2 {
```
Comment (leafty) [inline cloner/cmd/clone.go:91]: Why does this rule exist? Just check if there is a last part, i.e. `len() > 0`.

Comment (sgaist) [reply cloner/cmd/clone.go:91]: The standard repo URL for providers such as GitHub and GitLab have at least two components:
- namespace or group name (e.g. username, enterprise name, etc.)
- project name

for GitLab, you can also have subgroups.
This check avoids trying to clone wrongly formatted URL. Less ressources and time are wasted.

Comment (leafty) [reply cloner/cmd/clone.go:91]: But what if I want to clone a self-hosted git repository at `https://example.org/my-repo.git`? This check seems restrictive AND superfluous. If the URL is wrong, then `git clone` will fail, there is no benefit from not allowing it to fail.

Comment (sgaist) [reply cloner/cmd/clone.go:91]: I see you point.

The check has been updated

File: cloner/cmd/clone.go [hunk 3]
```diff
@@ -78,20 +84,44 @@ func clone(cmd *cobra.Command, args []string) {
...
 	}
 
+	if !strategy.Equal("default") {
+		_, err := os.Stat(clonePath)
+
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			log.Fatal("unexpected error: ", err)
+		}
+
+		if err == nil {
```
Comment (aledegano) [inline cloner/cmd/clone.go:109]: [I am out of my turf, but I am interested nonetheless]
This check is not necessary no? If `err != nil` we already exited with RC 1 in the block above. For readability I would rather remove this `if`.

File: cloner/cmd/clone.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sgaist) [reply cloner/cmd/clone.go]: I have refactored this part a bit to make things clearer with better naming and additional comments.

File: cloner/cmd/clone.go (hunk 3 above)
Comment (aledegano) [reply cloner/cmd/clone.go:119]: Right, sorry I got confused (and that is my point): too many nesting levels and negations for me to follow.

I think that I would rewrite this block as following:
```go
_, err := os.Stat(clonePath)

if err != nil && !errors.Is(err, os.ErrNotExist) {
  log.Fatal("unexpected error: ", err)
}

if errors.Is(err, os.ErrNotExist) {
  clonePathExists := true
}

if preCloningStrategy.Equal(NotIfExist) && clonePathExists {
  log.Print(clonePath, " already exist, doing nothing")
  os.Exit(0)
}

if preCloningStrategy.Equal(Overwrite)  && clonePathExists {
  log.Print("deleting clone")
  err = os.RemoveAll(clonePath + "/")
  if err != nil {
	log.Fatal("failed to remove existing clone: ", err)
  }
}
 
# Bonus log message, just for clarity
if preCloningStrategy.Equal(NoStrategy)  && clonePathExists {
 log.Print(clonePath, " already exists, but no pre-cloning strategy selected: git will handle this.")
}
```

Would you find this more readable? I would prefer it, but I will let you decide.

Comment (sgaist) [reply cloner/cmd/clone.go:119]: Your code is wrong on several places:
```
if errors.Is(err, os.ErrNotExist) {
  clonePathExists := true
}
```
should rather be: 
```
if errors.Is(err, os.ErrNotExist) {
  clonePathExists := false
}
```
but even so, clonePathExists lifetime is only within this if block so you will have two errors:
- Unused variable within that if block
- Undefined variable outside that if block

For the bonus log message, whether clonePathExists is true or false does not matter, the goal of this one is to just do nothing.

Anyway, your suggestion gave me a new idea on the implementation that I think should be satisfactory.

Comment (aledegano) [reply cloner/cmd/clone.go]: Yes, sorry the code was sort of a braindump. But I see that you understood what I meant and I find the new implementation much much easier to follow!
Thank you!

**Review by olevski — APPROVED** [review]

**Pull request merged**